server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -639,8 +642,11 @@ public String toString() {
         private final StoreFileMetaData md;
         private final Supplier<Integer> translogOps;
         private long position = 0;
+        private final Semaphore sendingTickets;
+        private volatile Exception error;
```

Comment:
perhaps an `AtomicReference<Exception>` and then further down below just do `error.compareAndSet(null, (Exception) ex);`